linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -3951,6 +3982,24 @@ public void close() {
     }
   }
 
+  private static class CartesianProductJoinEnumerator<TResult, TOuter, TInner>
+      extends CartesianProductEnumerator<Object, TResult> {
+    private final Function2<TOuter, TInner, TResult> resultSelector;
+
+    CartesianProductJoinEnumerator(Function2<TOuter, TInner, TResult> resultSelector,
+                                   Enumerator<Object> outer, Enumerator<Object> inner) {
```

Comment:
Should it be typed like `Enumerator<TOuter>`?

Reply:
Not possible if we want to extend `CartesianProductEnumerator`:
```
/**
 * Enumerator over the cartesian product of enumerators.
 * @param <T> Input element type
 * @param <E> Element type
 */
public abstract class CartesianProductEnumerator<T, E> implements Enumerator<E> {...}
```

Reply:
Well, can you cast when calling the constructor?

Reply:
Modified: cast in constructor